File: ios/Classes/CaptureQueue.swift
import Foundation

/// Converted PCM shared by the conversion and platform-channel workers.
struct CapturePacket {
    let generation, sequence, firstFrame, captureTimeNs: Int64
    let discontinuity: Bool
    let sampleRate: Double
    let frameCount: Int
    let audioData: Data
}

final class CaptureQueue {
    /// About two seconds of blocks, so a UI-isolate stall does not overflow.
    static func capacity(sampleRate: Double, frames: Int) -> Int {
        max(32, Int((2.0 * sampleRate / Double(frames)).rounded(.up)))
    }

    private let available = NSCondition()
    private let storage: UnsafeMutablePointer<Float>
    private let frames: Int
    private let capacity: Int
    private var positions: [Int64]
    private var times: [Int64]
    private var sequences: [Int64]
    private var gaps: [Bool]
    private var read = 0
    private var count = 0
    private var closed = false
    private var failure: Error?
    // Producer-only state.
    private var sequence: Int64 = 0
    private var gap = false

    init(frames: Int, capacity: Int) {
        self.frames = frames
        self.capacity = capacity
        positions = [Int64](repeating: 0, count: capacity)
        times = [Int64](repeating: 0, count: capacity)
        sequences = [Int64](repeating: 0, count: capacity)
        gaps = [Bool](repeating: false, count: capacity)
        storage = .allocate(capacity: frames * capacity)
        storage.initialize(repeating: 0, count: frames * capacity)
    }
    deinit {
        storage.deinitialize(count: frames * capacity)
        storage.deallocate()
    }

    /// `samples` holds exactly one block of `frames`.
    func offer(_ samples: UnsafePointer<Float>, position: Int64, timeNs: Int64, discontinuity: Bool) {
        let number = sequence
        sequence += 1
        available.lock()
        defer { available.unlock() }
        guard !closed else { return }
        if count == capacity { read = 0; count = 0; gap = true }
        let index = (read + count) % capacity
        storage.advanced(by: index * frames).update(from: samples, count: frames)
        positions[index] = position
        times[index] = timeNs
        sequences[index] = number
        gaps[index] = gap || discontinuity
        gap = false
        count += 1
        available.signal()
    }

    func take(generation: Int64, rate: Double, maximum: Int) throws -> [CapturePacket] {
        available.lock()
        defer { available.unlock() }
        let deadline = Date(timeIntervalSinceNow: 0.25)
        while count == 0 && !closed {
            if !available.wait(until: deadline) { break }
        }
        if let error = failure { throw error }
        var result = [CapturePacket]()
        for _ in 0..<min(count, max(1, min(8, maximum))) {
            let index = read
            let data = Data(bytes: storage.advanced(by: index * frames), count: frames * MemoryLayout<Float>.size)
            result.append(CapturePacket(generation: generation, sequence: sequences[index],
                firstFrame: positions[index], captureTimeNs: times[index],
                discontinuity: gaps[index], sampleRate: rate, frameCount: frames, audioData: data))
            read = (read + 1) % capacity
            count -= 1
        }
        return result
    }

    func close(_ error: Error? = nil) {
        available.lock()
        closed = true
        failure = error
        count = 0
        available.broadcast()
        available.unlock()
    }
}

File: android/src/main/kotlin/com/ymd/flutter_audio_capture/CaptureQueue.kt
package com.ymd.flutter_audio_capture

import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock
import kotlin.math.ceil

/** Fixed storage shared by the AudioRecord worker and bounded channel reads. */
internal class CaptureQueue(private val framesPerBlock: Int, capacity: Int) {
    companion object {
        /** About two seconds of blocks, so a UI-isolate stall does not overflow. */
        fun capacity(sampleRate: Int, framesPerBlock: Int) =
            maxOf(32, ceil(2.0 * sampleRate / framesPerBlock).toInt())
    }

    private val lock = ReentrantLock()
    private val available = lock.newCondition()
    private val samples = Array(capacity) { FloatArray(framesPerBlock) }
    private val positions = LongArray(capacity)
    private val times = LongArray(capacity)
    private val sequences = LongArray(capacity)
    private val gaps = BooleanArray(capacity)
    private var read = 0
    private var size = 0
    private var nextSequence = 0L
    private var gap = false
    private var closed = false
    private var failure: Throwable? = null

    fun offer(source: FloatArray, position: Long, timeNs: Long) {
        val sequence = nextSequence++
        // This is a blocking-read worker, not a real-time audio callback.
        // Contention with a bounded read must not discard valid PCM.
        lock.lock()
        try {
            if (closed) return
            if (size == samples.size) {
                read = 0
                size = 0
                gap = true
            }
            val index = (read + size) % samples.size
            source.copyInto(samples[index], endIndex = framesPerBlock)
            positions[index] = position
            times[index] = timeNs
            sequences[index] = sequence
            gaps[index] = gap
            gap = false
            size++
            available.signal()
        } finally { lock.unlock() }
    }

    fun take(generation: Long, rate: Int, maxBlocks: Int): List<Map<String, Any>> = lock.withLock {
        var remaining = 250_000_000L
        while (size == 0 && !closed && remaining > 0) remaining = available.awaitNanos(remaining)
        failure?.let { throw it }
        val result = ArrayList<Map<String, Any>>()
        repeat(minOf(size, maxBlocks)) {
            val index = read
            result.add(mapOf(
                "generation" to generation, "sequence" to sequences[index],
                "firstFrame" to positions[index], "captureTimeNs" to times[index],
                "discontinuity" to gaps[index],
                "sampleRate" to rate, "frameCount" to framesPerBlock, "audioData" to samples[index].copyOf()
            ))
            read = (read + 1) % samples.size
            size--
        }
        result
    }

    /** Marks the next offered block discontinuous, e.g. after an input route change. */
    fun markGap() = lock.withLock { gap = true }

    fun close(error: Throwable? = null) = lock.withLock {
        closed = true
        failure = error
        size = 0
        available.signalAll()
    }
}
